File: arithmetic_expression_evaluation.js
'use strict'
function toSuffixExperssion (infix) {
	let result = [], operators = []
	infix.forEach(item => {
		if(!isNaN(parseInt(item))) {
			result.push(item)
		} else {
			if ('()'.indexOf(item) > -1 || isHighest(item, operators)) {
				if(item === ')') {
					const lengthOfOperators = operators.length
					for(let i = 0; i < lengthOfOperators; i++) {
						const operator = operators.pop()
						if(operator !== '(') {
							result.push(operator)
						} else break
					}
				} else {
					operators.push(item)
				}
			} else { // 开始出栈并输出，直到最高为止。如果碰到了(,结束
				for(let i = operators.length - 1; i >= 0; i--) {
					if (operators[i] === '(' || isHighest(item, operators)) {
						operators.push(item)
						break
					} else {
						result.push(operators.pop())
					}
				}
				operators.length === 0 && operators.push(item)
			}
		}
	})
	return operators.length > 0 ? [...result, ...operators.reverse()] : result
}

/*
 * () ，在栈顶，可以压栈。开始出栈时又不能出栈
 */
function isHighest (operator, operators) {
	return (operators.length === 0 || getPriority(operator) > getPriority(operators[operators.length - 1])) 
}

/*
 * 未处理非法字符
 */
function getPriority (operator) {
	if (operator === '(') return -1
	return '+-'.indexOf(operator) > -1 ? 0 : 1
}


function getResult (input) {
	let tempStack = [], num1 = 0, num2 = 0
	const suffixExpression = toSuffixExperssion(input.split(' '))
  suffixExpression.forEach(item => {
  	if (!isNaN(parseInt(item))) {
  		tempStack.push(item)
  	} else {
  		num2 = parseInt(tempStack.pop())
  		num1 = parseInt(tempStack.pop())
  		tempStack.push(caculate(num1, num2, item))
  	}
  })
	return tempStack.length === 1 ? tempStack.pop() : NaN
}

function caculate (num1, num2, operator) {
	switch(operator) {
		case '+': return num1 + num2
		case '-': return num1 - num2
		case '*': return num1 * num2
		case '/': return num1 / num2
	}
	return 0
}

console.log(getResult('9 + ( 3 - 1 + 5 * 6 ) * 3 + 10 / 2'))

File: arithmetic_expression_evaluation.py
def toSuffixExpression (infix):
	result = []
	operators = []
	for item in infix:
		print(item, 'current result: ', result, infix)
		if item.isdigit():
			result.append(item)
		else:
			if '()'.find(item) > -1 or isHighest(item, operators):
				if item == ')':
					lengthOfOperators = len(operators)
					for i in range(lengthOfOperators):
						operator = operators.pop()
						if operator != '(':
							result.append(operator)
						else:
							break
				else:
					operators.append(item)
			else:
				lengthOfOperators = len(operators)
				for i in reversed(range(lengthOfOperators)):
					if operators[i] == '(' or isHighest(item, operators):
						operators.append(item)
						break
					else:
						result.append(operators.pop())
				return len(operators) == 0 and operators.append(item)
	return result + list(reversed(operators)) if len(operators) > 0 else result

def isHighest (operator, operators):
	result = len(operators) == 0 or getPriority(operator) > getPriority(operators[len(operators) - 1])
	print(operator, operators, '....highest:', result)
	return result

def getPriority (operator):
	if operator == '(':
		return -1
	return 0 if '+-'.find(operator) > -1 else 1

def getResult (input):
	tempStack = []
	num1 = 0
	num2 = 0
	suffixExpression = toSuffixExpression(input.split(' '))
	for item in suffixExpression:
		if item.isdigit():
			tempStack.append(item)
		else:
			num2 = int(tempStack.pop())
			num1 = int(tempStack.pop())
			tempStack.append(caculate(num1, num2, item))
	return tempStack.pop() if len(tempStack) == 1 else 0

	def caculate (num1, num2, operator):
		if operator == '+':
			return num1 + num2
		elif operator == '-':
			return num1 - num2
		elif operator == '*':
			return num1 * num2
		elif operator == '/':
			return num1 / num2
		return 0

print('the AEE for python is', toSuffixExpression('9 + ( 3 - 1 + 5 * 6 ) * 3 + 10 / 2'.split(' ')))












